### PR TITLE
feat: Prunable shard specs for streaming published segments

### PR DESCRIPTION
**Pull request by abhishekrb19**
Relates to https://github.com/apache/druid/issues/12929

  Streaming-published segments currently have numbered shard specs, which aren't prunable by design. Compaction must reindex the data with range or hashed partition strategy once the data is handed off — even if the topic is partitioned, which
  is easy to do with multiple supervisors. For multi-tenant datasources this means every tenant-filtered query hits every recent segment regardless of the partitioning strategy for numbered shard specs.

  This PR lets streaming tasks record, per published segment, the distinct values observed for a configured set of dimensions, and declare them on a new shard spec so the broker can prune near-realtime data without waiting for compaction to
  reindex handed off segments. Concurrent compaction cannot always keep up with the incoming data and additionally the compaction process itself takes time to reindex; so the benefits of range or hashed shard specs may not be fully realized for
  however long it takes to reindex (30-45 minutes in our case), and doesn't help with high concurrent query workloads that are only querying more recent data.

  So this PR adds a way to publish prunable shards right off the bat when they're handed off by streaming tasks, if configured. This functionality is opt-in, Kafka-only, and disabled by default.

  **Design**

  - `DimensionValueSetShardSpec` (type: `"dim_value_set"`) — extends `NumberedShardSpec` (behaves as a normal append segment) plus a partitionDimensionValues map (dimension → observed values). possibleInDomain() prunes a segment when the query
  constrains a declared dimension and none of its values intersect the domain; a dimension not in partitionDimensionValues is never pruned on. Set-based (not min/max), so it prunes precisely for sparse values and tolerates overlapping value
  sets across tasks/restarts.
  - Because partition values are observed at ingestion rather than hardcoded, incorrect or abruptly-changing partitioning never breaks correctness — at worst it yields non-prunable shard specs (similar to the default numbered shard) or bloated
  shard specs (this caveat can be addressed with a guard rail noted below).
  - Ingestion (SeekableStreamIndexTaskRunner) — when `partitionDimensions` is set, the task accumulates observed values per segment and stamps each at publish.

  **Configuration**

  New optional `tuningConfig` field `streamingPartitionsSpec` on the Kafka supervisor/task (default null), with a `partitionDimensions` list. When unset, behavior is unchanged. Documented in docs/ingestion/kafka-ingestion.md.

  **Compatibility**

  Backward-compatible and opt-in. But dim_value_set is a new core ShardSpec type with no defaultImpl fallback, so it is not forward-compatible: upgrade all services before enabling streamingPartitionsSpec, and note that once dim_value_set
  segments are published, downgrade isn't supported until they're compacted away.

  **Results**

  Tested in a cluster, where I saw up to ~40% reduction in segment scans on the historicals for a few low to medium cardinal partition dimensions. In a follow-up, I want to extend this to also prune tasks, for reduced peon buffers and better
  query performance at the task layer.

  **Caveats**

  There's currently no limit on the number of observed values stamped into a segment's partitionDimensionValues. It may make sense, in a follow-up, to add a configurable guardrail that falls back to NumberedShardSpec when the count exceeds a
  threshold, so shard specs don't get bloated.

  Release note

  Kafka ingestion can now publish segments that the broker prunes at query time, without waiting for compaction. Set tuningConfig.streamingPartitionsSpec.partitionDimensions to a list of low-to-medium cardinality dimensions; each task records
  the distinct values it observes per dimension and stamps them onto a new dim_value_set shard spec. Queries that filter on a declared dimension then skip segments whose values can't match. The feature is opt-in, Kafka-only, and disabled by
  default; when unset, behavior is unchanged.

  Compatibility: `dim_value_set` is a new core shard spec type with no fallback, so it is not forward-compatible. Upgrade all services before enabling streamingPartitionsSpec. Once dim_value_set segments are published, downgrade is unsupported
  until they are compacted away or streamingPartitionsSpec is removed.

This PR has:

- [x] been self-reviewed.
- [x] added documentation for new or modified features or behaviors.
- [x] a release note entry in the PR description.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [x] added integration tests.
- [x] been tested in a test Druid cluster.

### CODE REVIEW

File: indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamIndexTaskRunner.java [hunk 1]
```diff
@@ -999,15 +1057,86 @@ private void checkPublishAndHandoffFailure() throws ExecutionException, Interrup
...
+    );
+  }
+
   private void publishAndRegisterHandoff(SequenceMetadata<PartitionIdType, SequenceOffsetType> sequenceMetadata)
   {
     log.debug("Publishing segments for sequence [%s].", sequenceMetadata);
 
+    // annotateSegmentWithPartitionFilters is a no-op (returns the segment unchanged) when partition filters are not
+    // configured, so it is always safe to apply here.
+    final java.util.function.Function<Set<DataSegment>, Set<DataSegment>> shardSpecAnnotator =
```
Comment (aho135) [inline indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamIndexTaskRunner.java]: nit: can add Function to imports

File: indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamIndexTaskRunner.java [hunk 2]
```diff
@@ -1042,6 +1171,12 @@ public void onSuccess(SegmentsAndCommitMetadata publishedSegmentsAndCommitMetada
             );
             log.infoSegments(publishedSegmentsAndCommitMetadata.getSegments(), "Published segments");
 
+            for (DataSegment segment : publishedSegmentsAndCommitMetadata.getSegments()) {
+              observedDimensionValuesBySegment.remove(
+                  SegmentIdWithShardSpec.fromDataSegment(segment).toString()
```
Comment (aho135) [inline indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamIndexTaskRunner.java]: Should we also clean up `restartSpannedSegments.remove` here?

File: indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamIndexTaskRunner.java [hunk 3]
```diff
@@ -999,15 +1057,86 @@ private void checkPublishAndHandoffFailure() throws ExecutionException, Interrup
...
+    final String lookupKey = SegmentIdWithShardSpec.fromDataSegment(s).toString();
+    if (restartSpannedSegments.contains(lookupKey)) {
+      return s;
+    }
+    final Map<String, Set<String>> segObserved = observedDimensionValuesBySegment.get(lookupKey);
+    if (segObserved == null || segObserved.isEmpty()) {
+      return s;
+    }
+    final Map<String, List<String>> snapshotFilters = new HashMap<>();
+    for (String dim : filterDims) {
```
Comment (aho135) [inline indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamIndexTaskRunner.java]:
```
for (String dim : filterDims) {
    segObserved.computeIfPresent(dim, (k, vals) -> {
      synchronized (vals) {
        if (!vals.isEmpty()) {
          snapshotFilters.put(dim, new ArrayList<>(vals));
        }
      }
      return vals;  // Return unchanged - we're just reading
    });
  }
```

Comment (aho135) [reply indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamIndexTaskRunner.java]: Claude recommendation for race condition

File: docs/ingestion/kafka-ingestion.md [hunk 4]
```diff
@@ -126,6 +126,7 @@ For configuration properties shared across all streaming ingestion methods, refe
 |`pollTimeout`|Long|The length of time to wait for the Kafka consumer to poll records, in milliseconds.|No|100|
 |`useEarliestOffset`|Boolean|If a supervisor is managing a datasource for the first time, it obtains a set of starting offsets from Kafka. This flag determines whether the supervisor retrieves the earliest or latest offsets in Kafka. Under normal circumstances, subsequent tasks start from where the previous segments ended so this flag is only used on the first run.|No|`false`|
 |`idleConfig`|Object|Defines how and when the Kafka supervisor can become idle. See [Idle configuration](#idle-configuration) for more details.|No|null|
+|`partitionFilterDimensions`|List of String|Dimensions to track for query-time segment pruning. See [Partition filter dimensions](#partition-filter-dimensions) for details.|No|null|
```
Comment (aho135) [inline docs/ingestion/kafka-ingestion.md]: What do you think about naming this `partitionDimensions` to align with the compaction config? That may make it more clear that those values should be in sync

File: docs/ingestion/kafka-ingestion.md [hunk 5]
```diff
@@ -263,6 +264,46 @@ The following example shows a supervisor spec with idle configuration enabled:
...
+
+This enables segment pruning for streaming-ingested data without waiting for compaction to produce hash or range-partitioned segments.
+
+**Usage guidelines:**
+
+- Use only low-to-medium cardinality dimensions (for example, `tenant_id`, `region`, `environment`). High-cardinality dimensions bloat segment metadata with no pruning benefit.
+- Most effective when Kafka partitions are keyed by the tracked dimension (for example, using tenant ID as the message key). Each task naturally sees a subset of values, and segments get tight filter annotations.
+- Also works with multiple supervisors reading from separate topics into one datasource.
+- After compaction, the `StreamRangeShardSpec` annotations are replaced by the compaction output's shard spec (hash or range partitioning), which provides its own pruning.
```
Comment (aho135) [inline docs/ingestion/kafka-ingestion.md]: Maybe worth mentioning that when using `partitionFilterDimensions`, dynamic compaction strategy should not be used

File: indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamIndexTaskRunner.java [hunk 6]
```diff
@@ -999,15 +1057,86 @@ private void checkPublishAndHandoffFailure() throws ExecutionException, Interrup
...
+   */
+  @VisibleForTesting
+  DataSegment annotateSegmentWithPartitionFilters(DataSegment s)
+  {
+    final List<String> filterDims = ioConfig.getPartitionFilterDimensions();
+    if (CollectionUtils.isNullOrEmpty(filterDims)) {
+      return s;
+    }
+    final String lookupKey = SegmentIdWithShardSpec.fromDataSegment(s).toString();
+    if (restartSpannedSegments.contains(lookupKey)) {
```
Comment (FrankChen021) [inline indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamIndexTaskRunner.java]: P1 Mixed shard specs can fail publish after restart

Restart-spanned segments return unchanged as NumberedShardSpec, while new same-interval segments in the same publish batch can be annotated as StreamRangeShardSpec. TransactionalSegmentPublisher then runs SegmentPublisherHelper.annotateShardSpec, which rejects mixed shard-spec classes per interval, so a restarted task can fail publish/handoff. Make the fallback interval-wide, or stamp restored segments with a non-pruning stream_range spec.

File: extensions-core/kafka-indexing-service/src/main/java/org/apache/druid/indexing/kafka/supervisor/KafkaSupervisorIOConfig.java [hunk 7]
```diff
@@ -55,6 +57,62 @@ public class KafkaSupervisorIOConfig extends SeekableStreamSupervisorIOConfig
...
+        lateMessageRejectionPeriod,
+        earlyMessageRejectionPeriod,
+        lateMessageRejectionStartDateTime,
+        configOverrides,
+        idleConfig,
+        stopTaskCount,
+        emitTimeLagMetrics,
+        serverPriorityToReplicas,
+        boundedStreamConfig,
+        null
```
Comment (FrankChen021) [inline extensions-core/kafka-indexing-service/src/main/java/org/apache/druid/indexing/kafka/supervisor/KafkaSupervisorIOConfig.java]: P3 Backfill specs drop partitionFilterDimensions

This compatibility constructor always forwards null for partitionFilterDimensions. KafkaSupervisorSpec.createBackfillSpec still uses this overload when deriving bounded backfill specs, so a supervisor configured with partitionFilterDimensions silently creates backfill tasks without the pruning annotations. Pass the existing dimension list through for backfill specs.

**Review by FrankChen021 — COMMENTED** [review]
| Severity | Findings |
| --- | ---: |
| P0 | 0 |
| P1 | 1 |
| P2 | 0 |
| P3 | 1 |
| Total | 2 |

Reviewed 16 of 16 changed files.

I found two issues: restart-spanning segments can mix shard spec classes within one publish interval and fail publishing, and Kafka backfill specs drop partitionFilterDimensions.

---

This is an automated review by Codex GPT-5.5

**Review by FrankChen021 — COMMENTED** [review]
| Severity | Findings |
| --- | ---: |
| P0 | 0 |
| P1 | 0 |
| P2 | 0 |
| P3 | 1 |
| Total | 1 |

Reviewed 19 of 19 changed files.

---

This is an automated review by Codex GPT-5.5

File: indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamIndexTaskRunner.java [hunk 8]
```diff
@@ -999,15 +1060,80 @@ private void checkPublishAndHandoffFailure() throws ExecutionException, Interrup
...
+        if (vals == null) {
+          continue;
+        }
+        // vals is a synchronized set written by the run loop; copy it under its monitor to iterate safely.
+        final List<String> snapshot;
+        synchronized (vals) {
+          if (vals.isEmpty()) {
+            continue;
+          }
+          snapshot = new ArrayList<>(vals);
```
Comment (FrankChen021) [inline indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamIndexTaskRunner.java:1111]: [P3] Sort observed partition values before publishing

The list stored in partitionDimensionValues is created directly from a HashSet, so its order is unspecified. The embedded test already asserts a concrete order for these values, and equivalent published segment metadata can vary by JVM or run even when the value set is identical. Sort the snapshot deterministically, with explicit null handling, before putting it into the shard spec.

Reviewed 19 of 19 changed files.

Comment (abhishekrb19) [reply indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamIndexTaskRunner.java:1111]: Updated

File: docs/ingestion/kafka-ingestion.md (hunk 4 above)
Comment (abhishekrb19) [reply docs/ingestion/kafka-ingestion.md]: Yes, good call - thanks for the suggestion! I also moved this into the `tuningConfig` for consistency. It's nested under a `streamingPartitionsSpec` container to avoid naming ambiguity and so we can extend it with additional properties and/or add new types in the future if needed.

File: indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamIndexTaskRunner.java (hunk 6 above)
Comment (abhishekrb19) [reply indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamIndexTaskRunner.java]: Thanks, this is fixed with some test coverage added.

File: docs/ingestion/kafka-ingestion.md [hunk 9]
```diff
@@ -263,6 +263,50 @@ The following example shows a supervisor spec with idle configuration enabled:
...
+      "type": "kafka",
+      "topic": "events",
+      "consumerProperties": {"bootstrap.servers": "localhost:9092"},
+      "inputFormat": {"type": "json"},
+      "taskCount": 4,
+      "taskDuration": "PT1H"
+    },
+    "tuningConfig": {
+      "type": "kafka",
+      "streamingPartitionsSpec": {"partitionDimensions": ["tenant"]}
```
Comment (aho135) [inline docs/ingestion/kafka-ingestion.md:303]: Nice! Makes sense to have this in tuningConfig, and in the future we can add cardinality guardrails into `streamingPartitionsSpec` as well

File: extensions-core/kafka-indexing-service/src/test/java/org/apache/druid/indexing/kafka/KafkaIndexTaskTuningConfigTest.java [hunk 10]
```diff
@@ -128,6 +130,36 @@ public void testSerdeWithNonDefaults() throws Exception
...
+        TuningConfig.class
+    );
+
+    Assert.assertEquals(
+        new StreamingPartitionsSpec(List.of("tenant", "region")),
+        config.getStreamingPartitionsSpec()
+    );
+    Assert.assertEquals(List.of("tenant", "region"), config.getStreamingPartitionsSpec().getPartitionDimensions());
+  }
+
```
Comment (aho135) [inline extensions-core/kafka-indexing-service/src/test/java/org/apache/druid/indexing/kafka/KafkaIndexTaskTuningConfigTest.java:154]: Might be worth having a test case which covers edge case configuration for partitionDimensions (e.g. null values, empty string, integers)

File: indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamIndexTaskRunner.java [hunk 11]
```diff
@@ -999,15 +1062,82 @@ private void checkPublishAndHandoffFailure() throws ExecutionException, Interrup
...
+        }
+        // Sort for deterministic published metadata; null (missing value) sorts first.
+        snapshot.sort(Comparator.nullsFirst(Comparator.naturalOrder()));
+        snapshotFilters.put(dim, snapshot);
+      }
+    }
```
Comment (aho135) [inline indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamIndexTaskRunner.java:1117]: For a follow up, it may be useful to emit metric here for the cardinality of `observedPartitionDimValuesBySegment.get(segmentId)`

File: indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamIndexTaskRunner.java [hunk 12]
```diff
@@ -693,6 +735,27 @@ public void run()
...
+                      // Empty getDimension result means a null/missing value; record null so IS NULL is not pruned
+                      // (distinct from "", which getDimension returns as ["" ]).
+                      final List<String> dimValues = row.getDimension(dim);
+                      if (dimValues == null || dimValues.isEmpty()) {
+                        dimSet.add(null);
+                      } else {
+                        dimSet.addAll(dimValues);
+                      }
+                    }
+                  }
```
Comment (clintropolis) [inline indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamIndexTaskRunner.java]: it would be cool if there was an interface to abstract this 'do something to collect information for the shard spec' operation (thinking about other useful kinds of shard-specs like this, such as more classic ranges where we just collect min/max values, or if we could build bloom filters or something for higher cardinality columns).

I guess this would need streaming partition spec to either be an interface or have a 'processor' thingy that could provide the appropriate interface implementation.

This comment is not necessarily a blocker for this PR, just thinking how we could make something like this a bit more flexible/less specific for future improvements.

File: server/src/main/java/org/apache/druid/segment/realtime/appenderator/StreamAppenderatorDriver.java [hunk 13]
```diff
@@ -279,6 +279,16 @@ public ListenableFuture<SegmentsAndCommitMetadata> publish(
...
   )
+  {
+    return publish(publisher, committer, sequenceNames, java.util.function.Function.identity());
+  }
+
+  public ListenableFuture<SegmentsAndCommitMetadata> publish(
+      final TransactionalSegmentPublisher publisher,
+      final Committer committer,
+      final Collection<String> sequenceNames,
+      final java.util.function.Function<Set<DataSegment>, Set<DataSegment>> segmentAnnotateFunction
```
Comment (aho135) [inline server/src/main/java/org/apache/druid/segment/realtime/appenderator/StreamAppenderatorDriver.java:290]: `java.util.function.Function` can also be imported since it's used twice

**Review by aho135 — APPROVED** [review]
Left a few minor comments, but otherwise LGTM. Thanks @abhishekrb19!

File: server/src/main/java/org/apache/druid/segment/realtime/appenderator/StreamAppenderatorDriver.java (hunk 13 above)
Comment (abhishekrb19) [reply server/src/main/java/org/apache/druid/segment/realtime/appenderator/StreamAppenderatorDriver.java:290]: Yeah, I tried using Guava's `Function` here (which is used in the class below), but it doesn't seem to fit, hence the inline `java.util.function.Function`

**Review by FrankChen021 — COMMENTED** [review]
Reviewed 5 of 20 changed files (6 files total inspected) for this reply-needed follow-up.

Confirmed the two existing review follow-ups are addressed in current code: observed partition values are now sorted deterministically with explicit null handling, and restart-spanned segments now use empty-filter `StreamRangeShardSpec` behavior with relevant task-runner and publisher-helper test coverage.

No inline reply looks necessary.

---

This is an automated review by Codex GPT-5.5

File: extensions-core/kafka-indexing-service/src/test/java/org/apache/druid/indexing/kafka/KafkaIndexTaskTuningConfigTest.java (hunk 10 above)
Comment (abhishekrb19) [reply extensions-core/kafka-indexing-service/src/test/java/org/apache/druid/indexing/kafka/KafkaIndexTaskTuningConfigTest.java:154]: >Might be worth having a test case which covers edge case configuration for partitionDimensions (e.g. null values, empty string

I've added some tests in this unit test and there's some existing end-to-end coverage in the embedded ones too for null, empty cases

Re numeric types, they're currently unsupported for the `range` shard spec as well, given how the filters work - some details here: https://github.com/apache/druid/pull/19415. For now it's just documented for this new shard spec, but it would be nice to expand that functionality generally so all shard specs benefit from it; if that's not trivial, we could block them at creation time at least for this new shard spec in the future

**Comment by abhishekrb19** [comment]
Thanks for the reviews @aho135 @FrankChen021! I also went ahead and rename the shard spec type from `stream_range ` to `dim_value_set` as it better aligns with what it does.

File: indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamIndexTaskRunner.java (hunk 12 above)
Comment (abhishekrb19) [reply indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamIndexTaskRunner.java]: Yeah agreed, I think that should be doable. We could spec out the specifics of the interface contract sometime. Thanks for taking a look!
